lib/util/traffic-switch/traffic-switcher.ts

```diff
@@ -0,0 +1,139 @@
+  ExperimentName, aliasA and aliasB are unique identifiers and will be used for logging purposes.
+ */
+export abstract class TrafficSwitcher<T> {
+  public static readonly METRIC_NAME_TEMPLATE: string = 'TRAFFIC_SWITCHER__{EXP}__{METHOD}__{METRIC}'
+
+  // Used for logging purposes.
+  protected readonly experimentName: string
+  protected readonly aliasA: string
+  protected readonly aliasB: string
+  // 2 implementations of the interface that we want to switch traffic between.
+  protected readonly implementationA: T
+  protected readonly implementationB: T
+
+  // pctTrafficSwitchToB is the percentage of traffic that will be switched to implementationB
+  protected readonly pctTrafficSwitchToB: number
+  // pctSamplingToCompare is the percentage of traffic that will be compared between implementationA and implementationB
+  protected readonly pctSamplingToCompare: number
+
+  constructor(
+    experimentName: string,
+    implementationA: T,
+    implementationB: T,
+    aliasA: string,
+    aliasB: string,
+    pctTrafficSwitchToB: number = 0.0,
+    pctSamplingToCompare: number = 0.0
+  ) {
+    if (pctTrafficSwitchToB < 0 || pctTrafficSwitchToB > 1) {
+      throw new Error('Percentages must be between 0 and 1')
+    }
+    if (pctTrafficSwitchToB > 0 && pctSamplingToCompare > 0) {
+      throw new Error('Only one of pctTrafficSwitchToB and pctSamplingToCompare can be greater than 0')
+    }
+
+    this.experimentName = experimentName
+    this.implementationA = implementationA
+    this.implementationB = implementationB
+    this.aliasA = aliasA
+    this.aliasB = aliasB
+    this.pctTrafficSwitchToB = pctTrafficSwitchToB
+    this.pctSamplingToCompare = pctSamplingToCompare
+  }
+
+  /* This method will do one of the following:
+   *  - Switch traffic between two implementations based on the pctTrafficSwitchToB percentage
+   *  - Sample traffic between two implementations based on the pctSamplingToCompare percentage
+   */
+  protected async trafficSwitchMethod<K>(
+    methodA: () => Promise<K>,
+    methodB: () => Promise<K>,
+    methodName: string,
+    defaultReturnVal: K,
+    comparisonMethod: ((resultA: K | undefined, resultB: K | undefined) => void) | undefined
+  ): Promise<K> {
+    if (this.isSamplingEnabled()) {
+      if (this.shouldSample()) {
+        this.logMetric(methodName, `COMPARISON_SAMPLE`)
+        // If in sampling mode, and we should sample, call both implementations and compare results
+        // Note: here we'll wait for both implementations to finish before comparing results.
+        const results: (K | undefined)[] = await this.allSettled<K>([
+          [methodA(), this.aliasA, methodName],
+          [methodB(), this.aliasB, methodName],
+        ])
+
+        // Compare the results if a comparison method is provided
+        if (results && results.length == 2) {
+          try {
+            if (comparisonMethod) {
+              comparisonMethod(results[0], results[1])
+            }
```

Comment:
`comparisonMethod && comparisongMethod(results[0], results[1])` could also work if you like that